### PR TITLE
#958 fixing tedge mapper tests

### LINKED ISSUE #958: Observation: Message on tedge/errors is recently missing when we are not connected
**Describe the bug**

This is an observation about functionality that has recently changed and that was detected by two system-tests.
Not sure if the tests needs to be updated or else.

Usually, when we caused a parsing error e.g. with this: `sudo tedge mqtt pub tedge/measurements {` then we received a complaint on topic `tedge/errors` when tedge-mapper-c8y was running.

Recently, it looks like we really need to connect to C8y to get the same error message on the error topic.

**To Reproduce**

These two tests were testing the behavior:
```
pysys.py run tedge_mapper_c8y_positive
pysys.py run tedge_mapper_c8y_negative
```

**Expected behavior**

Test should pass or be fixed.
This could be achieved by subclassing the test to EnvironmentC8y. 

**Screenshots**

First approach. We connect before causing an error:

sudo tedge connect c8y
sudo tedge mqtt pub tedge/measurements {
image

On MQTT:
image

Second approach: we only start the tege-mapper-c8y service and then cause the error (no 

sudo systemctl start tedge-mapper-c8y.service
sudo tedge mqtt pub tedge/measurements {
image

On MQTT:
image

### PR DESCRIPTION
**Pull request by initard**
## Proposed changes
  - removed tedge_mapper_c8y_negative
  - removed tedge_mapper_c8y_positive
  - added the same tests in rust using mqtt_tests

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
-  #958 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by PradeepKiruvale — APPROVED** [review]
Very nice!. Thanks for porting to Rust.

**Comment by abelikt** [comment]
Very cool,
as far as I understand this test creates http mocks for C8y and the JWT token magic.
Then it spawns a mapper in tokio and then exchanges some messages.
It would be cool to have a short description somewhere (e.g. in line 136 and 142).

I see this test also as a base for other tests. It would be possible to move the mocking to a separate (setup)-function so that we can easily create other tests that send and recieve messages.

Do you use internal iterfaces or would it be possible to make it an iteration
test in the mapper crate?
E.g. my moving it here:
crates/core/tedge_mapper/test

File: crates/core/tedge_mapper/src/c8y/mapper.rs [hunk 1]
```diff
@@ -108,3 +108,105 @@ fn create_directories() -> Result<(), anyhow::Error> {
...
+    const DEVICE_NAME: &str = "test-user";
+    const DEVICE_TYPE: &str = "test-thin-edge.io";
+    const MQTT_HOST: &str = "127.0.0.1";
+
+    #[test_case(
+        "tedge/measurements",
+        "tedge/errors",
+        "{",
+        "Invalid JSON: EOF while parsing an object at line 1 column 1: `{`"
+    )] // fail case
```
Comment (abelikt) [inline crates/core/tedge_mapper/src/c8y/mapper.rs:136]: Maybe here a small descriptive sentence

File: crates/core/tedge_mapper/src/c8y/mapper.rs [hunk 2]
```diff
@@ -108,3 +108,105 @@ fn create_directories() -> Result<(), anyhow::Error> {
...
+    #[test_case(
+        "tedge/measurements",
+        "tedge/measurements",
+        r#"{"temperature": 12, "time": "2021-06-15T17:01:15.806181503+02:00"}"#,
+        r#"{"temperature": 12, "time": "2021-06-15T17:01:15.806181503+02:00"}"#
+    )] // successs case
```
Comment (abelikt) [inline crates/core/tedge_mapper/src/c8y/mapper.rs:142]: Maybe here a small descriptive sentence, too

File: crates/core/tedge_mapper/src/c8y/mapper.rs [hunk 3]
```diff
@@ -108,3 +108,105 @@ fn create_directories() -> Result<(), anyhow::Error> {
...
+                json!({ "externalId": DEVICE_ID, "managedObject": { "id": "123" } }).to_string(),
+            )
+            .create();
+
+        let mut jwt_token_retriver = Box::new(MockC8yJwtTokenRetriever::new());
+        jwt_token_retriver
+            .expect_get_jwt_token()
+            .returning(|| Ok(SmartRestJwtResponse::default()));
+
+        let http_client = reqwest::ClientBuilder::new().build().unwrap();
```
Comment (abelikt) [inline crates/core/tedge_mapper/src/c8y/mapper.rs:163]: I see only http here, is it using https in the background? I'm just curious how we could manage and test https  requests.

File: crates/core/tedge_mapper/src/c8y/mapper.rs [hunk 4]
```diff
@@ -108,3 +108,105 @@ fn create_directories() -> Result<(), anyhow::Error> {
...
+                .await
+                .unwrap();
+        });
+
+        // publish `payload` to `pub_topic`
+        let () = broker.publish(pub_topic, payload).await?;
+
+        // check the `messages` returned contain `expected_msg`
+        assert_received_all_expected(&mut messages, TEST_TIMEOUT_SECS, &[expected_msg]).await;
+        Ok(())
```
Comment (abelikt) [inline crates/core/tedge_mapper/src/c8y/mapper.rs:210]: I hope that everything that could fail here will panic

**Review by abelikt — APPROVED** [review]
Approved with some minor comments, thanks!

**Comment by initard** [comment]
> Do you use internal iterfaces or would it be possible to make it an iteration
> test in the mapper crate?

Sadly yes, Internal interfaces are used

File: crates/core/tedge_mapper/src/c8y/mapper.rs (hunk 4 above)
Comment (initard) [reply crates/core/tedge_mapper/src/c8y/mapper.rs:210]: Yes, this is part of the Rust test crate

File: crates/core/tedge_mapper/src/c8y/mapper.rs (hunk 3 above)
Comment (didier-wenzek) [reply crates/core/tedge_mapper/src/c8y/mapper.rs:167]: You are correct, the tests are done using http not https. An issue like [this one](https://github.com/thin-edge/thin-edge.io/issues/1082) will not be detected by these tests.

**Review by didier-wenzek — COMMENTED** [review]
We have to be aware that these tests don't replace the previous ones. The former system tests were actually testing the final executable in a real context. These new tests are integration tests that exercise an ad-hoc combination of structs (`JwtAuthHttpProxy` + `CumulocityConverter` + `Mapper`). This improves code coverage - compared to unit tests of the `CumulocityConverter`. But these tests will be very fragile faced to refactoring.

**Pull request merged**